### ISSUE TITLE
keycloak immutability and local instance config

### ISSUE THREAD

**Issue by mark-prins** [openmsupply/unified-codes#149]
Some issues need updating with the current state of the keycloak server:
- Have a specific admin user and password for a particular instance of keycloak
- the H2 database is cleared when the container is restarted and the config needs to be reloaded

## Expected Behaviour
The default admin user shouldn't be allowed to be admin/admin and the config should only be required on initialisation. 
Perhaps we need to run a database instance outside of the supplied docker container? 

## Current Behaviour
Currently the docker script sets the user as `admin` and password as `admin` - there needs to be a simple way to set the admin password on initialisation.
The config shouldn't have to be reloaded every time the container is started.

**Comment by wlthomson** [comment]
Have simplified issue title as it seems both these isuses are related to data not being persisted when the container is spun down.

I think this is generally intended behaviour, and agree with proposed solution. From my experience/what I've read, I think the standard way to fix this is to store the database outside of the container and map the container ports appropriately (so that nothing inside the container needs to be changed). Similar thing with the config?

**Comment by katherine-sussol** [comment]
Yeah agree with external DB idea. Dgraph works similarly in that it stores the persistent data on the file system and maps to the container using volumes (i.e. nothing persisted inside container). I think the idea is the containers are supposed to be transient (at least based on what I've also been reading). 

Not too sure about the password - I guess the main concern is that we currently have it hardcoded in the run script (which is publicly available?)

**Comment by mark-prins** [comment]
>  I guess the main concern is that we currently have it hardcoded in the run script

Yeah, that stood out when I reviewed. had me thinking that we could be more secure. Storing the db outside the container would allow this, as well as having a default password to get people up and running 😃

**Comment by wlthomson** [comment]
Some good documentation on this here: https://github.com/keycloak/keycloak-containers/blob/master/server/README.md.

Are we happy to stick with `h2`, or maybe go with something a bit more familiar to us like `postgres`?

**Comment by mark-prins** [comment]
I think `postgres` : I resisted because it seems a bit sledgehammery for our current use case, and we have a lot of servers spinning up already!

If we are running with another server, I'd go with something we know and are using in other places.

**Comment by wlthomson** [comment]
>  I think the idea is the containers are supposed to be transient...

Yeah, as far as I'm aware containers are just "instances" of images, anything that exists inside a container is gone when that container is stopped. The common solution AFAIK is to map the container storage to the host (see https://docs.docker.com/storage/volumes/), so that data can be persisted outside of the container and reloaded if the container is restarted (just as you described re. dgraph).
